### PR TITLE
report dropped attrs

### PR DESCRIPTION
**Pull request by burdettadam**
Signed-off-by: Adam Burdett <burdettadam@gmail.com>

### CODE REVIEW

File: aries_cloudagent/messaging/jsonld/create_verify_data.py [hunk 1]
```diff
@@ -35,59 +37,78 @@ def _cannonize_document(doc):
...
+    if len(data.get(attr[0], {})) > len(framed.get(attr[1], {})):
+        for_diff = jsonld.compact(expanded, data.get("@context"))
+        dropped = set(data.get(attr[0], {}).keys()) - set(
+            for_diff.get(attr[1], {}).keys()
+        )
+        raise DroppedAttributeException(
+            f"in {attr[0]}, {dropped} attributes dropped."
+            "Provide definitions in context to correct."
+        )
+        raise DroppedAttributeException("CredentialSubject dropped")
```
Comment (dbluhm) [inline aries_cloudagent/messaging/jsonld/create_verify_data.py:111]: Should be able to simplify this by iterating through a list that looks like:
```python
attrs = [("proof", "proof"), ("credentialSubject", "https://www.w3.org/....#credentialSubject")]
```

Setting intermediate values for the `get`s would also help with readability, I think. There's a lot of characters in there so having some of that unraveled a bit would make it easier to parse.

File: aries_cloudagent/messaging/jsonld/tests/test_credential.py [hunk 2]
```diff
@@ -329,6 +422,24 @@ async def test_sign_credential(input, mock_session):
     assert "jws" in result.get("proof", {}).keys()
 
 
+@pytest.mark.parametrize("input", TEST_SIGN_ERROR_OBJS)
+@pytest.mark.asyncio
+async def test_sign_dropped_attribute_exception(input, mock_session):
+    with pytest.raises(DroppedAttributeException):
+        await sign_credential(
+            mock_session, input.get("doc"), input.get("options"), TEST_VERKEY
```
Comment (dbluhm) [inline aries_cloudagent/messaging/jsonld/tests/test_credential.py:430]: Probably better to use `input["doc"]` over `get`. `doc` being absent is a stronger error condition than defaulting to `None` and the `None` wouldn't cause an error immediately. "Fail fast, fix faster" :wink:

File: aries_cloudagent/messaging/jsonld/create_verify_data.py (hunk 1 above)
Comment (burdettadam) [reply aries_cloudagent/messaging/jsonld/create_verify_data.py:111]: good idea, I will make the change. I also will remove the trailing exception on line 111.

File: aries_cloudagent/messaging/jsonld/create_verify_data.py [hunk 3]
```diff
@@ -35,59 +37,71 @@ def _cannonize_document(doc):
...
-    if (
-        "type" not in signature_options
-        or signature_options["type"] != "Ed25519Signature2018"
-    ):
-        signature_options["type"] = "Ed25519Signature2018"
+    signature_options["created"] = signature_options.get("created", created_at())
 
-    [expanded] = jsonld.expand(data)
+    # TODO: povide methods for jsonld expand that reports dropped attributes for better error reporting
+    [expanded] = jsonld.expand(data, {"keepFreeFloatingNodes": True})
```
Comment (dbluhm) [inline aries_cloudagent/messaging/jsonld/create_verify_data.py:75]: We should remove the `keepFreeFloatingNodes` we added while testing

File: aries_cloudagent/messaging/jsonld/create_verify_data.py [hunk 4]
```diff
@@ -35,59 +37,71 @@ def _cannonize_document(doc):
...
+        dropped = set(data.keys()) - set(for_diff.keys())
+        raise DroppedAttributeException(
+            f"{dropped} attributes dropped. "
+            " Provide definitions in context to correct."
+        )
+    # Check proof for dropped attributes
+    attr = [("proof", "proof"), ("credentialSubject",
+        "https://www.w3.org/2018/credentials#credentialSubject")]
+    data_context = data.get("@context")
+    for maping in attr:
```
Comment (dbluhm) [inline aries_cloudagent/messaging/jsonld/create_verify_data.py:92]: Mapping is spelled with two p's

File: aries_cloudagent/messaging/jsonld/tests/test_credential.py (hunk 2 above)
Comment (burdettadam) [reply aries_cloudagent/messaging/jsonld/tests/test_credential.py:461]: Of course! How could I forget the Five F's of Fix-It Felix development. "Fail Fast, Fix Faster and don't Fail!" Remember the Four R's of review, "Review, Remember, Refocus and Review!"

File: aries_cloudagent/messaging/jsonld/create_verify_data.py [hunk 5]
```diff
@@ -35,59 +37,71 @@ def _cannonize_document(doc):
...
+            "signature_options.verificationMethod is required"
         )
 
-    if "created" not in signature_options:
-        signature_options["created"] = datetime.datetime.now(
-            datetime.timezone.utc
-        ).strftime("%Y-%m-%dT%H:%M:%SZ")
+    def created_at():
+        stamp = datetime.datetime.now(datetime.timezone.utc)
+        return stamp.strftime("%Y-%m-%dT%H:%M:%SZ")
```
[the remark this answers is not in the document]
Comment (burdettadam) [reply aries_cloudagent/messaging/jsonld/create_verify_data.py]: Done!

**Review by dbluhm — APPROVED** [review]
Looks good!

**Comment by dbluhm** [comment]
Tests checked out, nice work!

**Pull request merged**